### ISSUE TITLE
OpenJ9 (IBM Semeru) and Java 17

### ISSUE THREAD

**Issue by zdegner** [microsoft/ApplicationInsights-Java#2016]
### Expected behavior
I'm working on a series of services with a common framework and have a few questions. The framework consists of Micronaut 3.2.2, AppInsights Java Agent 3.2.3, and a few other libraries. Right now the services are running with a Hotspot JVM (image: mcr.microsoft.com/openjdk/jdk:17-mariner) without any issues. However, when switching to OpenJ9 (image: ibm-semeru-runtimes:open-17-jdk) we get some curious errors about "illegally accessing 'private' member of class java/lang/reflect". These errors only appear when the AppInsights Agent is attached to the JVM, the services work fine without the agent.

I'm wondering if there are any suggestions on how to get around this error. I've tried adding the following without success:
--add-opens=java.base/java.lang.reflect=java.base
--add-opens=java.base/java.lang.reflect=ALL-UNNAMED

What is curious is that the error seems to imply illegal reflective access between the same module (java.base). If anyone has any ideas on how to debug this issue further it would be appreciated.

#### Rationale behind using OpenJ9:
We would like to switch back to OpenJ9 now that a Java 17 compatible release is available. Through some load testing against the Java 11+17 JVMs and inspection with NMT tool we have found that each of our Hotspot microservices requires at least 1.5 GB of memory assigned in Kubernetes (ideally more). The Hotspot JVM appears to slowly leak native memory and then eventually is OOMKilled under continuous load. OpenJ9 on the other hand is able to operate in 900 MB of memory without ever leaking native memory or exceeding its quota.

### Actual behavior
[dental-assessment-service] 01:49:12.159 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: Error instantiating bean of type [io.micronaut.flyway.$Alternati
veMigrationRunner$Definition$Intercepted]: Class java/lang/reflect/Proxy$ProxyBuilder$$Lambda$23/0x00000000311d4ac0 illegally accessing "private" member of class java/lang/reflect/Prox
y$ProxyBuilder
[dental-assessment-service] io.micronaut.context.exceptions.BeanInstantiationException: Error instantiating bean of type [io.micronaut.flyway.$AlternativeMigrationRunner$Definition$Int
ercepted]: Class java/lang/reflect/Proxy$ProxyBuilder$$Lambda$23/0x00000000311d4ac0 illegally accessing "private" member of class java/lang/reflect/Proxy$ProxyBuilder
[dental-assessment-service]     at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2365)
[dental-assessment-service]     at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:3281)
[dental-assessment-service]     at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:3267)
[dental-assessment-service]     at io.micronaut.context.DefaultBeanContext.initializeEventListeners(DefaultBeanContext.java:1838)
[dental-assessment-service]     at io.micronaut.context.DefaultBeanContext.readAllBeanDefinitionClasses(DefaultBeanContext.java:3452)
[dental-assessment-service]     at io.micronaut.context.DefaultBeanContext.finalizeConfiguration(DefaultBeanContext.java:3883)
[dental-assessment-service]     at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:329)
[dental-assessment-service]     at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:183)
[dental-assessment-service]     at io.micronaut.runtime.Micronaut.start(Micronaut.java:72)
...
[dental-assessment-service] Caused by: java.lang.IllegalAccessError: Class java/lang/reflect/Proxy$ProxyBuilder$$Lambda$23/0x00000000311d4ac0 illegally accessing "private" member of cl
ass java/lang/reflect/Proxy$ProxyBuilder
[dental-assessment-service]     at java.base/java.lang.reflect.Proxy$ProxyBuilder$$Lambda$23/0x00000000311d4ac0.apply(Unknown Source)
[dental-assessment-service]     at java.base/jdk.internal.loader.AbstractClassLoaderValue$Memoizer.get(AbstractClassLoaderValue.java:329)
[dental-assessment-service]     at java.base/jdk.internal.loader.AbstractClassLoaderValue.computeIfAbsent(AbstractClassLoaderValue.java:205)
[dental-assessment-service]     at java.base/java.lang.reflect.Proxy$ProxyBuilder.getDynamicModule(Proxy.java:908)
[dental-assessment-service]     at java.base/java.lang.reflect.Proxy$ProxyBuilder.mapToModule(Proxy.java:846)
[dental-assessment-service]     at java.base/java.lang.reflect.Proxy$ProxyBuilder.<init>(Proxy.java:651)
[dental-assessment-service]     at java.base/java.lang.reflect.Proxy$ProxyBuilder.<init>(Proxy.java:656)
[dental-assessment-service]     at java.base/java.lang.reflect.Proxy.lambda$getProxyConstructor$0(Proxy.java:429)
[dental-assessment-service]     at java.base/java.lang.reflect.Proxy$$Lambda$412/0x000000009023cbf0.apply(Unknown Source)
[dental-assessment-service]     at java.base/jdk.internal.loader.AbstractClassLoaderValue$Memoizer.get(AbstractClassLoaderValue.java:329)
[dental-assessment-service]     at java.base/jdk.internal.loader.AbstractClassLoaderValue.computeIfAbsent(AbstractClassLoaderValue.java:205)
[dental-assessment-service]     at java.base/java.lang.reflect.Proxy.getProxyConstructor(Proxy.java:427)
[dental-assessment-service]     at java.base/java.lang.reflect.Proxy.newProxyInstance(Proxy.java:1037)
[dental-assessment-service]     at java.base/sun.reflect.annotation.AnnotationParser$1.run(AnnotationParser.java:302)
[dental-assessment-service]     at java.base/sun.reflect.annotation.AnnotationParser$1.run(AnnotationParser.java:300)
[dental-assessment-service]     at java.base/java.security.AccessController.doPrivileged(AccessController.java:683)
[dental-assessment-service]     at java.base/sun.reflect.annotation.AnnotationParser.annotationForMap(AnnotationParser.java:300)
[dental-assessment-service]     at java.base/sun.reflect.annotation.AnnotationParser.parseAnnotation2(AnnotationParser.java:289)
[dental-assessment-service]     at java.base/sun.reflect.annotation.AnnotationParser.parseAnnotations2(AnnotationParser.java:121)
[dental-assessment-service]     at java.base/sun.reflect.annotation.AnnotationParser.parseAnnotations(AnnotationParser.java:73)
[dental-assessment-service]     at java.base/java.lang.Class.getAnnotationCache(Class.java:3105)
[dental-assessment-service]     at java.base/java.lang.Class.getAnnotation(Class.java:2690)
[dental-assessment-service]     at io.micronaut.core.annotation.AnnotationMetadata.isRepeatableAnnotation(AnnotationMetadata.java:1499)
[dental-assessment-service]     at io.micronaut.inject.annotation.DefaultAnnotationMetadata.isRepeatableAnnotation(DefaultAnnotationMetadata.java:1305)
[dental-assessment-service]     at io.micronaut.core.annotation.AnnotationMetadata.hasDeclaredStereotype(AnnotationMetadata.java:1464)
[dental-assessment-service]     at io.micronaut.core.annotation.AnnotationMetadataDelegate.hasDeclaredStereotype(AnnotationMetadataDelegate.java:514)
[dental-assessment-service]     at io.micronaut.inject.BeanDefinition.isIterable(BeanDefinition.java:165)
[dental-assessment-service]     at io.micronaut.context.DefaultApplicationContext.transformIterables(DefaultApplicationContext.java:258)
[dental-assessment-service]     at io.micronaut.context.DefaultApplicationContext.findBeanCandidates(DefaultApplicationContext.java:249)
[dental-assessment-service]     at io.micronaut.context.DefaultBeanContext.findConcreteCandidateNoCache(DefaultBeanContext.java:3090)
[dental-assessment-service]     at io.micronaut.context.DefaultBeanContext.findConcreteCandidate(DefaultBeanContext.java:3054)
[dental-assessment-service]     at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2750)
[dental-assessment-service]     at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1638)
[dental-assessment-service]     at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:1933)
[dental-assessment-service]     at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1189)
[dental-assessment-service]     at io.micronaut.flyway.$AlternativeMigrationRunner$Definition$Intercepted$Definition.build(Unknown Source)
[dental-assessment-service]     at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2336)
[dental-assessment-service]     ... 10 common frames omitted

### To Reproduce
I'm not sure yet what combination of libraries causes this issue. I may try to put together a Micronaut starter with OpenJ9 and AppInsights Java agent to see if that is enough.

### System information
applicationinsights-agent: 3.2.3
applicationinsights-core: 2.6.4
Java: ibm-semeru-runtimes:open-17-jdk (OpenJ9 0.29.1, OpenJDK 17)
Open Telemetry: 1.7.0
Micronaut: 3.2.2
Hibernate: 5.6.0.Final

**Comment by trask** [comment]
hi @zdegner! if you can provide a repro that would be really helpful, thx

**Comment by zdegner** [comment]
@trask I tried to create a reproducible project based off of the Micronaut starter (running it in a docker container). The AppInsights Java Agent worked fine with JDK 17, OpenJ9 and Micronaut. After systematically adding every library we use (Hibernate, Hikari CP, Redis, Azure SDK, etc) to the sample project I circled back to our framework and performed a Gradle "clean" before rebuilding the libraries and downstream services. The container now starts up normally.

I cannot really explain what went wrong here, but the builds appear to be working now. It could be some intermediate artifact that was cached -- or something bizarre like that. I think it is safe to close this issue though.
